pkg/api/cachedcsi.go

```diff
@@ -0,0 +1,203 @@
+	}
+
+	if req.VolumeCapability == nil {
+		return nil, status.Error(codes.InvalidArgument, "NodePublishVolume Volume Capability must be provided")
+	}
+
+	targetPath := req.GetTargetPath()
+	volumeID := req.GetVolumeId()
+	volumeAttributes := req.GetVolumeContext()
+
+	var cachePath string
+	var targetPermissions os.FileMode
+
+	if suffix, exists := volumeAttributes["placeCacheAtPath"]; exists {
+		// running in suffix mode, desired outcome:
+		//  - the mount point is writable by the pod
+		//  - the cache is mounted at the suffix, and is not writable
+		cachePath = path.Join(targetPath, suffix)
+		targetPermissions = 0777
+	} else {
+		// running in unsuffixed mode, desired outcome:
+		//  - the mount point *is* the cache, and is not writable by the pod
+		cachePath = targetPath
+		targetPermissions = 0755
+	}
+
+	if err := os.MkdirAll(targetPath, targetPermissions); err != nil {
+		return nil, fmt.Errorf("failed to create target directory %s: %s", targetPath, err)
+	}
+
+	if err := os.Chmod(targetPath, targetPermissions); err != nil {
+		return nil, fmt.Errorf("failed to change ownership of target directory %s: %s", targetPath, err)
+	}
+
+	version, err := c.WriteCache(cachePath)
+	if err != nil {
+		return nil, err
+	}
+
+	logger.Info(ctx, "volume published", key.VolumeID.Field(volumeID), key.TargetPath.Field(targetPath), key.Version.Field(version))
+
+	return &csi.NodePublishVolumeResponse{}, nil
+}
+
+func (s *Cached) NodeUnpublishVolume(ctx context.Context, req *csi.NodeUnpublishVolumeRequest) (*csi.NodeUnpublishVolumeResponse, error) {
+	if req.VolumeId == "" {
+		return nil, status.Error(codes.InvalidArgument, "NodeUnpublishVolume Volume ID must be provided")
+	}
+
+	if req.TargetPath == "" {
+		return nil, status.Error(codes.InvalidArgument, "NodeUnpublishVolume Target Path must be provided")
+	}
+
+	targetPath := req.GetTargetPath()
+
+	// Clean up directory
+	if err := os.RemoveAll(targetPath); err != nil {
+		return nil, fmt.Errorf("failed to remove directory %s: %s", targetPath, err)
+	}
+
+	logger.Info(ctx, "volume unpublished and data removed", key.TargetPath.Field(targetPath))
+	return &csi.NodeUnpublishVolumeResponse{}, nil
+}
+
+// NodeGetVolumeStats returns the volume capacity statistics available for the given volume.
+func (c *Cached) NodeGetVolumeStats(ctx context.Context, req *csi.NodeGetVolumeStatsRequest) (*csi.NodeGetVolumeStatsResponse, error) {
+	if req.VolumeId == "" {
+		return nil, status.Error(codes.InvalidArgument, "NodeGetVolumeStats Volume ID must be provided")
+	}
+
+	volumePath := req.VolumePath
+	if volumePath == "" {
+		return nil, status.Error(codes.InvalidArgument, "NodeGetVolumeStats Volume Path must be provided")
+	}
+
+	usedBytes, err := getFolderSize(volumePath)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to retrieve used size statistics for volume path %q: %s", volumePath, err)
```

Comment:
Why are you using `%q` for the path, according to the docs that's for a single character.

Reply:
*ai-generated shame*